ppy/osu-framework
# Pack native libs in .net core fashion

**Warning: This PR won't pass CI before merge**, unless adding a temporary workaround commit. You can do all the checks at local to **validate it does work**.

I've done the validations on Windows. Since there's no runtime gap unlike Mono comparing with .Net Framework, I except mac and Linux to simply success.

## Motivation
.NET Core introduces a better way to [configure native dependencies via package](https://docs.microsoft.com/en-us/nuget/create-packages/supporting-multiple-target-frameworks#architecture-specific-folders). Now legacy technologies get dropped and .NET Core is the only supported runtime for desktop, it's time to reform from various workarounds and get served by toolset conventions. The Xamarin builds ships totally different assets and won't be affected by this.

## Changes
The native libraries located in `runtimes/<rid>/native/` are well-handled by the dotnet sdk. No more need to manually setting path/naming conventions for them.

Calls to `SetDllDirectory` for Windows will be removed from framework.
Linux users no longer require to set the environment variable `LD_LIBRARY_PATH`.

`dotnet publish` will handle them properly. Deploy script should be updated to depend on it; old implications about file layout may break.
It will remove other OS's assets when packaging for specific OS. I can see `.dll` and `.so` in macOS package now.

Framework package changes to targeting both .NET Core and .NET Standard, and reference native libs package only for .NET Core target.
This can prevent miss-installing for the native package, as it's only designed for .NET Core.
Cross-framework targeting will be required further for framework, because .NET Core will be the only framework that's null-annotated in 1 year.

Size of framework NuGet package will be duped, because it contains assemblies for both framework. It's no harmful as the package is small enough today (~2.5M), and will only affect development machine.

## Validating
### Why requiring manual validation
Though most parts of `dotnet` toolset handle it well, **`ProjectReference` doesn't know about native dependencies.** It will copy the files to the correct layout, but missing `project.assets.json` and `.deps.json` entries for them. This will result in `DllNotFoundException` at runtime, because CLR won't just load anything inside `runtimes\<rid>\native` folder. It only load known dependencies.

I have spent a whole day investigating if I can inject something into `.deps.json`. The task creating it accepts some dependency lists, but they won't be used. It follows exactly `project.assets.json`, which is generated by one concrete task. Copying them directly aside executable will have other problems too.

Since native libs seldom get updated, and ships in different pace comparing to framework, I think the best solution is to **only consume it via package reference**.
If the package following new convention hasn't be published to NuGet gallery, it can't be consumed by CI since we don't have nightly feeds like .NET official. Thankfully, NuGet provides capability to test package consuming at local machine.

### Steps to validate
Based on the latest commit.
1. Create a package with higher version than stable.
```
cd osu.Framework.NativeLibs
dotnet pack -p:version=9999.9.9
```

2. Put the package into local feed
Assuming directory `~\.nuget\hook\`
```
dotnet nuget push .\bin\Debug\ppy.osu.Framework.NativeLibs.9999.9.9.nupkg -s ~\.nuget\hook\
```
or
```
nuget.exe push .\bin\Debug\ppy.osu.Framework.NativeLibs.9999.9.9.nupkg -Source ~\.nuget\hook\
```

3. Update the package reference in framework project to using the higher version, and restore it with local feed
You can it directly in Visual Studio's NuGet setting, or restore using
```
 dotnet restore -s ~\.nuget\hook\
```
It's recommended to close your IDE and clean all the `bin` and `obj` folder before restore to ensure they're clean.

4. Build SampleGame.Desktop and desktop tests.
Check `SampleGame.Desktop.deps.json`. Search for `bass.dll`, `libbass.so` and `libbass.dylib`, you should find entries for them.
If there are native assets out of `runtime` folder, that's directly put outside or in `x86`/`x64`/`x86-64` folder, remove them to avoid misleading.

5. Run the sample game and visual tests. You shouldn't suffer `DllNotFoundException`.

6. Switch to the game repo. Install the latest package using
```
cd osu.Desktop
dotnet add package -v 9999.9.9 -s ~\.nuget\hook\ ppy.osu.Framework.NativeLibs
```
Verify desktop game also works. The new package is compatible with old framework, because workarounds are simply no-op when CLR can locate them.

7. Verify `dotnet publish -r` also works.
Be aware that if runtime identifier are specified, native assets will be put out directly, not in `runtimes` folder.

8. Don't forget to clean your NuGet global cache at `~\.nuget\packages\ppy.osu.Framework.NativeLibs\9999.9.9`.
